### PR TITLE
Add support to optimize for NDCG at a given truncation level

### LINKED ISSUE #3420: LambdarankNDCG objective ignores truncation level
In /src/objective/rank_objective.hpp: currently, for each query, all pairs of documents contribute to the gradients computation regardless of the truncation level _k_. However, to correctly optimize for NDCG@_k_, one should exclude pairs containing both documents beyond the top _k_ (as they don't affect NDCG@_k_ when swapped). Please enforce the check for the truncation level _k_.

### PR DESCRIPTION
**Pull request by metpavel**
Fixes #3420

### CODE REVIEW

**Comment by guolinke** [comment]
Any benchmarks for the change?

**Comment by metpavel** [comment]
> Any benchmarks for the change?

For MSLR-WEB30K dataset (fold 1) using default train.conf, performance on the validation set:
Optimize for NDCG@FullRank:
ndcg@1 : 0.498897
ndcg@3 : 0.484631
ndcg@5 : 0.487863
208.163323 seconds elapsed, finished iteration 100  

Optimize for NDCG@20:
ndcg@1 : 0.502574
ndcg@3 : 0.486615
ndcg@5 : 0.489424
124.763394 seconds elapsed, finished iteration 100  

Optimize for NDCG@10:
ndcg@1 : 0.504389
ndcg@3 : 0.487059
ndcg@5 : 0.48924   
93.361508 seconds elapsed, finished iteration 100

Optimize for NDCG@5:
ndcg@1 : 0.498769
ndcg@3 : 0.483403
ndcg@5 : 0.48611
96.900527 seconds elapsed, finished iteration 100

Optimize for NDCG@3:
ndcg@1 : 0.503184
ndcg@3 : 0.483097
ndcg@5 : 0.485419
79.423862 seconds elapsed, finished iteration 100

Optimize for NDCG@1:
ndcg@1 : 0.502715
ndcg@3 : 0.47437
ndcg@5 : 0.47624
71.433821 seconds elapsed, finished iteration 100

**Comment by guolinke** [comment]
I think this PR is a new feature.

**Comment by guolinke** [comment]
@metpavel Is the result of "full rank" the same as before?

File: src/objective/rank_objective.hpp [hunk 1]
```diff
@@ -163,35 +163,30 @@ class LambdarankNDCG : public RankingObjective {
...
-      double high_sum_hessian = 0.0;
-      for (data_size_t j = 0; j < cnt; ++j) {
-        // skip same data
-        if (i == j) {
-          continue;
-        }
-        const data_size_t low = sorted_idx[j];
+    // start accmulate lambdas by pairs that contain at least one document above truncation level
+    for (data_size_t i = 0; i < cnt - 1 && i < truncation_level_; ++i) {
+      for (data_size_t j = i + 1; j < cnt; ++j) {
```
Comment (guolinke) [inline src/objective/rank_objective.hpp:168]: Isn't this start from '0' ?  if not, why?

Comment (guolinke) [inline src/objective/rank_objective.hpp:168]:
```suggestion
      if (score[sorted_idx[i]] == kMinScore) { continue; }
      for (data_size_t j = i + 1; j < cnt; ++j) {
        if (score[sorted_idx[j]] == kMinScore) { continue; }
```

File: src/objective/rank_objective.hpp [hunk 2]
```diff
@@ -163,35 +163,30 @@ class LambdarankNDCG : public RankingObjective {
...
         const double low_score = score[low];
-        // only consider pair with different label
-        if (high_label <= low_label || low_score == kMinScore) {
-          continue;
-        }
+        const double low_label_gain = label_gain_[low_label];
+        const double low_discount = DCGCalculator::GetDiscount(low_rank);
 
+        if (high_score == kMinScore || low_score == kMinScore) { continue; }
+
```
Comment (guolinke) [inline src/objective/rank_objective.hpp:187]:
```suggestion
```

File: src/objective/rank_objective.hpp [hunk 3]
```diff
@@ -163,35 +163,30 @@ class LambdarankNDCG : public RankingObjective {
...
+        if (label[sorted_idx[i]] == label[sorted_idx[j]]) { continue; }
+        
+        const data_size_t high_rank = label[sorted_idx[i]] > label[sorted_idx[j]] ? i : j;
+        const data_size_t high = sorted_idx[high_rank];
+        const int high_label = static_cast<int>(label[high]);
+        const double high_score = score[high];
+        const double high_label_gain = label_gain_[high_label];
+        const double high_discount = DCGCalculator::GetDiscount(high_rank);
+        
+        const data_size_t low_rank = label[sorted_idx[i]] > label[sorted_idx[j]] ? j : i;
```
Comment (guolinke) [inline src/objective/rank_objective.hpp:179]: remove the additional branching? get high_rank and low_rank by one if.

Comment (metpavel) [reply src/objective/rank_objective.hpp]: done

**Comment by metpavel** [comment]
> @metpavel Is the result of "full rank" the same as before?

@guolinke, yes

**Comment by guolinke** [comment]
From the results, it seems we need a slightly larger truncation level, e.g. k + 3, to optimize NDCG@k?

**Comment by metpavel** [comment]
> From the results, it seems we need a slightly larger truncation level, e.g. `k + m`, to optimize `NDCG@k`?

@guolinke, perhaps one may want to treat that _m_ as a tunable hyper-parameter to balance between better alignment with the desired cutoff _k_ (smaller m) and more pairs to train on (larger _m_)

**Comment by guolinke** [comment]
@metpavel Yes, maybe you can add this description into the document of truncation level?

**Comment by metpavel** [comment]
> @metpavel Yes, maybe you can add this description into the document of truncation level?

@guolinke, do you mean here: /blob/master/docs/Parameters.rst?

**Comment by guolinke** [comment]
@metpavel we usually writh the doc for parameters in `config.h`, and then run `helpers/parameter_generator.py` to auto update .rst file.

**Comment by guolinke** [comment]
BTW, can you also merge master, to ensure CI passes?

**Comment by guolinke** [comment]
@jameslamb can you help to fix the R's tests? It needs to update some ranking score constraints, as the ranking objective is updated.

**Comment by jameslamb** [comment]
> @jameslamb can you help to fix the R's tests? It needs to update some ranking score constraints, as the ranking objective is updated.

ok sure, I can help

**Comment by jameslamb** [comment]
@metpavel I just created a pull request into this branch on your fork. If you merge it, it should fix the R tests

https://github.com/metpavel/LightGBM/pull/1

File: src/objective/rank_objective.hpp [hunk 4]
```diff
@@ -208,16 +211,13 @@ class LambdarankNDCG : public RankingObjective {
...
+        hessians[high] += static_cast<score_t>(p_hessian);
         // lambda is negative, so use minus to accumulate
         sum_lambdas -= 2 * p_lambda;
       }
-      // update
-      lambdas[high] += static_cast<score_t>(high_sum_lambda);
-      hessians[high] += static_cast<score_t>(high_sum_hessian);
     }
     if (norm_ && sum_lambdas > 0) {
       double norm_factor = std::log2(1 + sum_lambdas) / sum_lambdas;
```
Comment (jameslamb) [inline src/objective/rank_objective.hpp:223]: /gha run-valgrind

Comment (jameslamb) [reply src/objective/rank_objective.hpp:223]: /gha run r-valgrind

Comment (jameslamb) [reply src/objective/rank_objective.hpp:223]: ^ this comment started this run checking that these changes pass our valgrind tests: https://github.com/microsoft/LightGBM/runs/1296338580?check_suite_focus=true

Comment (jameslamb) [reply src/objective/rank_objective.hpp:223]: this didn't show any new issues...but I'll run it again once my fixes are merged into this PR: https://github.com/microsoft/LightGBM/pull/3425#issuecomment-714915235

File: include/LightGBM/config.h [hunk 5]
```diff
@@ -835,7 +835,8 @@ struct Config {
 
   // check = >0
   // desc = used only in ``lambdarank`` application
-  // desc = used for truncating the max DCG, refer to "truncation level" in the Sec. 3 of `LambdaMART paper <https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/MSR-TR-2010-82.pdf>`__
+  // desc = controls the number of top-results to focus on during training, refer to "truncation level" in the Sec. 3 of `LambdaMART paper <https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/MSR-TR-2010-82.pdf>`
+  // desc = is closely related to the desirable cutoff k in the metric NDCG@k that we aim at optimizing the ranker for. The optimal setting for this parameter is likely to be slightly higher than k (e.g., k + 3) to include more pairs of documents to train on, but perhaps not too high to avoid deviating too much from the desired target metric NDCG@k
   int lambdarank_truncation_level = 20;
```
Comment (guolinke) [inline include/LightGBM/config.h:840]:
```suggestion
  int lambdarank_truncation_level = 30;
```

Comment (guolinke) [reply include/LightGBM/config.h:840]: I think increase the default value, may have better default performance.

**Comment by guolinke** [comment]
You need to re-run the parameter generator, due to the default value is changed.

**Review by guolinke — APPROVED** [review]

File: src/objective/rank_objective.hpp (hunk 4 above)
Comment (jameslamb) [reply src/objective/rank_objective.hpp:219]: /gha run-valgrind

Comment (jameslamb) [reply src/objective/rank_objective.hpp:219]: /gha run r-valgrind

Comment (jameslamb) [reply src/objective/rank_objective.hpp:219]: if this passes, I'll come back and approve: https://github.com/microsoft/LightGBM/runs/1310275649?check_suite_focus=true
